File: downloader/lib/publisher.py
"""Publish a scheduled post row to Instagram via Outstand."""

from datetime import datetime, timezone
from pathlib import Path

from lib import db, outstand


def publish_post_row(post: dict) -> dict:
    """Given a joined row from db.due_posts / db.list_posts, push it live.

    Mutates the posts table: status, outstand_post_id, outstand_media_id, posted_at, error_message.
    """
    post_id = post["id"]
    local_path = Path(post["local_path"])
    caption = (post.get("caption") or "").strip()
    account_id = (post.get("outstand_account_id") or "").strip()

    if not caption:
        _fail(post_id, "no caption attached")
        raise RuntimeError(f"post {post_id}: no caption")
    if not account_id:
        _fail(post_id, "no outstand_account_id")
        raise RuntimeError(f"post {post_id}: no outstand_account_id")
    if not local_path.exists():
        _fail(post_id, f"video file missing: {local_path}")
        raise RuntimeError(f"post {post_id}: video file missing: {local_path}")

    db.update_post(post_id, status="publishing", error_message=None)

    try:
        media = outstand.upload_local_video(local_path, filename_base=f"interior-{post_id}")
        result = outstand.create_post(
            accounts=[account_id],
            caption=caption,
            media_url=media["url"],
            filename=media["filename"],
        )
    except Exception as e:
        _fail(post_id, str(e))
        raise

    db.update_post(
        post_id,
        status="posted",
        outstand_post_id=result["post_id"],
        outstand_media_id=media["media_id"],
        posted_at=datetime.now(timezone.utc).isoformat(),
        error_message=None,
    )
    db.mark_video_used(post["video_id"])

    return {"post_id": post_id, "outstand_post_id": result["post_id"], "status": result["status"]}


def _fail(post_id: int, msg: str) -> None:
    db.update_post(post_id, status="failed", error_message=msg[:500])

File: lib/publisher.ts
import "server-only";

import { and, eq, lte } from "drizzle-orm";

import { db, captions, posts, videos } from "@/lib/db";
import { createPost, uploadRemoteVideo } from "@/lib/outstand";

export type PublishResult = {
  postId: number;
  outstandPostId?: string;
  status: "posted" | "failed" | "skipped";
  error?: string;
};

/** Publish every post whose scheduled_at has passed. Returns per-post results. */
export async function publishDuePosts(): Promise<PublishResult[]> {
  const now = new Date();

  const duePosts = await db
    .select({
      postId: posts.id,
      videoId: posts.videoId,
      captionId: posts.captionId,
      outstandAccountId: posts.outstandAccountId,
      status: posts.status,
      cloudinaryUrl: videos.cloudinaryUrl,
      title: videos.title,
      captionText: captions.text,
    })
    .from(posts)
    .innerJoin(videos, eq(posts.videoId, videos.id))
    .leftJoin(captions, eq(posts.captionId, captions.id))
    .where(and(eq(posts.status, "scheduled"), lte(posts.scheduledAt, now)));

  const results: PublishResult[] = [];

  for (const row of duePosts) {
    if (!row.captionText || !row.captionText.trim()) {
      await markFailed(row.postId, "no caption attached");
      results.push({ postId: row.postId, status: "failed", error: "no caption" });
      continue;
    }
    if (!row.cloudinaryUrl) {
      await markFailed(row.postId, "no cloudinary url on video");
      results.push({ postId: row.postId, status: "failed", error: "no cloudinary url" });
      continue;
    }

    await db
      .update(posts)
      .set({ status: "publishing", errorMessage: null })
      .where(eq(posts.id, row.postId));

    try {
      const media = await uploadRemoteVideo({
        sourceUrl: row.cloudinaryUrl,
        filenameBase: `interior-${row.postId}`,
      });
      const result = await createPost({
        accounts: [row.outstandAccountId],
        caption: row.captionText,
        mediaUrl: media.url,
        filename: media.filename,
      });

      await db
        .update(posts)
        .set({
          status: "posted",
          outstandPostId: result.postId,
          outstandMediaId: media.mediaId,
          postedAt: new Date(),
        })
        .where(eq(posts.id, row.postId));

      await db
        .update(videos)
        .set({ usedForPost: 1 })
        .where(eq(videos.id, row.videoId));

      results.push({ postId: row.postId, outstandPostId: result.postId, status: "posted" });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      await markFailed(row.postId, message);
      results.push({ postId: row.postId, status: "failed", error: message });
    }
  }

  return results;
}

async function markFailed(postId: number, message: string) {
  await db
    .update(posts)
    .set({ status: "failed", errorMessage: message.slice(0, 500) })
    .where(eq(posts.id, postId));
}
